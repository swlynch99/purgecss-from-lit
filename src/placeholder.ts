import { createHash } from 'crypto';

// In a separate file so it can be imported by both index.ts and its tests.
export const placeholder = 'placeholder' + createHash('sha256')
    .update('placeholder')
    .digest('hex');
