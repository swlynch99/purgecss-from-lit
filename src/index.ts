import { TSESTreeOptions, parse } from '@typescript-eslint/typescript-estree';
import { walk } from 'estree-walker';
import { createHash } from 'crypto';
import parseHtml, { ExtractorResultDetailed } from 'purgecss-from-html';

import { placeholder } from './placeholder';

export type { ExtractorResultDetailed } from 'purgecss-from-html';
export type { TSESTreeOptions } from '@typescript-eslint/typescript-estree';

export type PluginOptions = TSESTreeOptions;

type PluginFunction = (content: string) => ExtractorResultDetailed;

const defaultOptions: TSESTreeOptions = {
  allowInvalidAST: true,
}

/**
 * Create a function to extract selectors from lit web components.
 * 
 * This will look for selectors within html templates used by lit. It is not
 * able to find values inserted dynamically into the template.
 * 
 * @param options - `@typescript-eslint/typescript-estree` options.
 * @returns a function to extract selectors from lit web components.
 */
export default function purgeFromLit(options?: TSESTreeOptions): PluginFunction {
  return purgeFromLitImpl({
    ...defaultOptions,
    ...(options || {})
  });
}

function purgeFromLitImpl(options: TSESTreeOptions) {
  return (content: string): ExtractorResultDetailed => {
    const ast = parse(content, options);

    const attributes = new Set<string[]>();
    const classes = new Set<string>();
    const ids = new Set<string>();
    const tags = new Set<string>();
    const undetermined = new Set<string>();

    walk(ast as any, {
      enter(node): void {
        if (node.type !== "TaggedTemplateExpression")
          return;

        if (node.tag.type !== "Identifier")
          return;
        if (node.tag.name !== "html")
          return;

        const fragments = node.quasi.quasis.map(quasi => quasi.value.cooked!);
        const result = parseHtml(fragments.join(placeholder));

        extendSet(classes, result.classes);
        extendSet(ids, result.ids);
        extendSet(tags, result.tags);
        extendSet(undetermined, result.undetermined);

        const { names, values } = result.attributes;
        for (let [name, value] of names.map((name, i) => [name, values[i]])) {
          // Ignore lit properties and event listeners. They don't correspond
          // to real HTML attributes and so shouldn't be included.
          if (name.startsWith('.') || name.startsWith('@'))
            continue;

          // For boolean attributes we transform them to look as if the
          // attribute is just present. (i.e. `?disabled=${cond}` is changed to
          // just be `disabled`).
          //
          // The value should always be a template here so this should be fine.
          if (name.startsWith('?')) {
            name = name.substring(1);
            value = ''
          }

          if (name === placeholder) {
            classes.add(value);
            ids.add(value);
          }

          attributes.add([name, value]);
        }
      },
    })

    return {
      attributes: unzipAttributes(attributes),
      classes: [...classes],
      ids: [...ids],
      tags: [...tags],
      undetermined: [...undetermined]
    }

    throw "not implemented yet"
  }
}

function extendSet<T>(set: Set<T>, values: T[]) {
  for (const value of values) {
    set.add(value);
  }
}

function unzipAttributes(attrs: Iterable<string[]>) {
  let names = [];
  let values = [];

  for (const [name, value] of attrs) {
    names.push(name);
    values.push(value);
  }

  return { names, values }
}
