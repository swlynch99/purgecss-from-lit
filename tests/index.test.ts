import { expect, test } from 'vitest';
import { createHash } from 'crypto';

import purgeFromLit from '../src/index';
import { placeholder } from '../src/placeholder';

const plugin = purgeFromLit();

test('lit property template value', () => {
    expect(plugin('html`<div class=${thang}></div>`'))
        .toMatchSnapshot()
})

test('template with attribute', () => {
    expect(plugin('html`<div class="blah" disabled data-tooltip=${tooltip}></div>`'))
        .toMatchSnapshot()
})

test('template within a template', () => {
    const result = plugin(`
        html\`
            <div class="outer" data-tooltip=\${tooltip}>
                \${html\`<span class=\${debug}>inner span</span>\`}
            </div>
        \`
    `);

    expect(result.classes).includes("outer");
    expect(result.classes).includes(placeholder);
    expect(result.attributes.names).includes("data-tooltip");
})

test('template boolean attribute', () => {
    const result = plugin(`
        html\`<span ?hidden=\${value}></span>\`
    `);

    expect(result.attributes.names).to.deep.equal(['hidden']);
    expect(result.attributes.values).to.deep.equal(['']);
})

test('template with property attribute', () => {
    const result = plugin(`
        html\`<span .value=\${value}></span>\`
    `)

    expect(result.attributes.names).to.be.empty;
    expect(result.attributes.values).to.be.empty;
})

test('template with an event listener', () => {
    const result = plugin(`
        html\`
            <button @click=\${event => event.preventDefault()}>
                Click Here
            </button>
        \`
    `)

    expect(result.attributes.names).to.be.empty
    expect(result.attributes.values).to.be.empty
})

test('template with HTML comment', () => {
    const result = plugin(`
        html\`
            <!-- A comment with an inner <div class=\${expression}> -->
        \`
    `)

    expect(result.attributes.names).to.be.empty
})

test('custom element types shows up as a potential class', () => {
    const result = plugin(`
        html\`<div \${prop}="thonk"></div>\`
    `)

    expect(result.classes).to.contain("thonk")
})

test('multiple HTML elements', () => {
    const result = plugin(`
        const prop = inline\`class\`;
        function doTheThink(event: Thonk) {}
    
        function thingy() {
            return html\`
                <div class="main">
                    <button @click=\${doTheThink}>Do the think!</button>
                    <custom-component .prop="5" \${prop}="thonk"></custom-component>
                </div>
            \`
        }
    `)

    expect(result).toMatchSnapshot()
})

test('HTML outside of a literal is not included', () => {
    const result = plugin(`
        const a = "<div class=\\"main\\"></div>";
        const b = inline\`<a href=\${a}>lonk</a>\`;
    `)

    expect(result.classes).to.be.empty
    expect(result.ids).to.be.empty
    expect(result.tags).to.be.empty
})
