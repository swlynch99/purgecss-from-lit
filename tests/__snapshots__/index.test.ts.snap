// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lit property template value 1`] = `
{
  "attributes": {
    "names": [],
    "values": [],
  },
  "classes": [
    "placeholder4097889236a2af26c293033feb964c4cf118c0224e0d063fec0a89e9d0569ef2",
  ],
  "ids": [],
  "tags": [
    "head",
    "div",
    "body",
    "html",
  ],
  "undetermined": [],
}
`;

exports[`multiple HTML elements 1`] = `
{
  "attributes": {
    "names": [
      "placeholder4097889236a2af26c293033feb964c4cf118c0224e0d063fec0a89e9d0569ef2",
    ],
    "values": [
      "thonk",
    ],
  },
  "classes": [
    "main",
    "thonk",
  ],
  "ids": [
    "thonk",
  ],
  "tags": [
    "head",
    "button",
    "custom-component",
    "div",
    "body",
    "html",
  ],
  "undetermined": [],
}
`;

exports[`template with attribute 1`] = `
{
  "attributes": {
    "names": [
      "disabled",
      "data-tooltip",
    ],
    "values": [
      "",
      "placeholder4097889236a2af26c293033feb964c4cf118c0224e0d063fec0a89e9d0569ef2",
    ],
  },
  "classes": [
    "blah",
  ],
  "ids": [],
  "tags": [
    "head",
    "div",
    "body",
    "html",
  ],
  "undetermined": [],
}
`;
